### PR TITLE
compiler: Patch placement of ConditionalDimension with multi-Dimension conditions

### LINKED ISSUE #2007: Conditional dimension depending on muliple dimensions placed in innermost loop
Consider this proxy for a Fourier integral:
```
from devito import (TimeFunction, ConditionalDimension, Function,
                    DefaultDimension, Grid, Eq, Operator)
from sympy import Ge
import numpy as np

freq = [2,4,8]
nfreq = np.shape(freq)[0]
grid = Grid(shape=(5,5))
# time function for holding wavefield proxy
u = TimeFunction(name="u", grid=grid)
t = grid.time_dim
# frequency dimension for holding fft proxy
freq_dim = DefaultDimension(name='freq_dim', default_value=nfreq)
f = Function(name="f", grid=grid, dimensions=(freq_dim,), shape=(nfreq,))
# load the frequencies into f, this will serve as a lower bound for the proxy integral
f.data[:] = np.array(freq[:])
# proxy fourier integral holder
ure = Function(name="ure", grid=grid, dimensions=(freq_dim,) + u.indices[1:], shape=(nfreq,) + u.shape[1:])
# make update for u
u_up = Eq(u.forward, u+1)
# conditional dimension based on f to simulate bounds of fourier integral
ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=True)
# proxy for fourier integral
fourier_eq = Eq(ure, ure + u, implicit_dims=ct)
op = Operator([u_up, fourier_eq])
op.apply(time_M=10)

print(ure.data)

```

It returns the following correct result:
```
[[[54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]
  [54. 54. 54. 54. 54.]]

 [[49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]
  [49. 49. 49. 49. 49.]]

 [[27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]
  [27. 27. 27. 27. 27.]]]
```

However, if we examine the C code we find that the conditional dimension acting as the integral bound has been placed within the innermost loop:
```
#define _POSIX_C_SOURCE 200809L
#define MIN(a,b) (((a) < (b)) ? (a) : (b))
#define MAX(a,b) (((a) > (b)) ? (a) : (b))
#define START_TIMER(S) struct timeval start_ ## S , end_ ## S ; gettimeofday(&start_ ## S , NULL);
#define STOP_TIMER(S,T) gettimeofday(&end_ ## S, NULL); T->S += (double)(end_ ## S .tv_sec-start_ ## S.tv_sec)+(double)(end_ ## S .tv_usec-start_ ## S .tv_usec)/1000000;

#include "stdlib.h"
#include "math.h"
#include "sys/time.h"
#include "xmmintrin.h"
#include "pmmintrin.h"

struct dataobj
{
  void *restrict data;
  unsigned long * size;
  unsigned long * npsize;
  unsigned long * dsize;
  int * hsize;
  int * hofs;
  int * oofs;
  void * dmap;
} ;

struct profiler
{
  double section0;
  double section1;
} ;


int Kernel(struct dataobj *restrict f_vec, struct dataobj *restrict u_vec, struct dataobj *restrict ure_vec, const int freq_dim0_blk0_size, const int freq_dim_M, const int freq_dim_m, const int time_M, const int time_m, const int x0_blk0_size, const int x_M, const int x_m, const int y_M, const int y_m, struct profiler * timers)
{
  float (*restrict f) __attribute__ ((aligned (64))) = (float (*)) f_vec->data;
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]]) u_vec->data;
  float (*restrict ure)[ure_vec->size[1]][ure_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[ure_vec->size[1]][ure_vec->size[2]]) ure_vec->data;

  /* Flush denormal numbers to zero in hardware */
  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);
  _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);

  for (int time = time_m, t0 = (time)%(2), t1 = (time + 1)%(2); time <= time_M; time += 1, t0 = (time)%(2), t1 = (time + 1)%(2))
  {
    /* Begin section0 */
    START_TIMER(section0)
    for (int x = x_m; x <= x_M; x += 1)
    {
      #pragma omp simd aligned(u:64)
      for (int y = y_m; y <= y_M; y += 1)
      {
        u[t1][x + 1][y + 1] = u[t0][x + 1][y + 1] + 1;
      }
    }
    STOP_TIMER(section0,timers)
    /* End section0 */

    /* Begin section1 */
    START_TIMER(section1)
    for (int freq_dim0_blk0 = freq_dim_m; freq_dim0_blk0 <= freq_dim_M; freq_dim0_blk0 += freq_dim0_blk0_size)
    {
      for (int x0_blk0 = x_m; x0_blk0 <= x_M; x0_blk0 += x0_blk0_size)
      {
        for (int freq_dim = freq_dim0_blk0; freq_dim <= MIN(freq_dim0_blk0 + freq_dim0_blk0_size - 1, freq_dim_M); freq_dim += 1)
        {
          for (int x = x0_blk0; x <= MIN(x0_blk0 + x0_blk0_size - 1, x_M); x += 1)
          {
            #pragma omp simd aligned(f,u,ure:64)
            for (int y = y_m; y <= y_M; y += 1)
            {
              if (time >= f[freq_dim])
              {
                ure[freq_dim][x + 1][y + 1] = u[t0][x + 1][y + 1] + ure[freq_dim][x + 1][y + 1];
              }
            }
          }
        }
      }
    }
    STOP_TIMER(section1,timers)
    /* End section1 */
  }

  return 0;
}
```

If I modify the conditional dimension definition to be the following:
`ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=False)`
I get the following compiler error:
```
gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c -lm -o /tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.so
/tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c: In function ‘Kernel’:
/tmp/devito-jitcache-uid1000/bb75d42a661d62a79aa80ab0ea52cddc47b82dff.c:56:19: error: ‘freq_dim’ undeclared (first use in this function); did you mean ‘freq_dim_m’?
   56 |     if (time >= f[freq_dim])
      |                   ^~~~~~~~
      |                   freq_dim_m
```

If I modify the conditional dimension and the implicit dimensions of the integral equation like this:
`ct = ConditionalDimension(name="ct", parent=t, condition=Ge(t,f), indirect=False)`
`fourier_eq = Eq(ure, ure + u, implicit_dims=(freq_dim, ct))`

It compiles but I get the following incorrect result:
```
[[[94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]
  [94. 94. 94. 94. 94.]]

 [[73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]
  [73. 73. 73. 73. 73.]]

 [[31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]
  [31. 31. 31. 31. 31.]]]
```

Examining the code, this is because a second time loop has been created following the updates to U, so the integral is only really updated based on U at the last time step:
```
#define _POSIX_C_SOURCE 200809L
#define START_TIMER(S) struct timeval start_ ## S , end_ ## S ; gettimeofday(&start_ ## S , NULL);
#define STOP_TIMER(S,T) gettimeofday(&end_ ## S, NULL); T->S += (double)(end_ ## S .tv_sec-start_ ## S.tv_sec)+(double)(end_ ## S .tv_usec-start_ ## S .tv_usec)/1000000;

#include "stdlib.h"
#include "math.h"
#include "sys/time.h"
#include "xmmintrin.h"
#include "pmmintrin.h"

struct dataobj
{
  void *restrict data;
  unsigned long * size;
  unsigned long * npsize;
  unsigned long * dsize;
  int * hsize;
  int * hofs;
  int * oofs;
  void * dmap;
} ;

struct profiler
{
  double section0;
  double section1;
} ;


int Kernel(struct dataobj *restrict f_vec, struct dataobj *restrict u_vec, struct dataobj *restrict ure_vec, const int freq_dim_M, const int freq_dim_m, const int time_M, const int time_m, const int x_M, const int x_m, const int y_M, const int y_m, struct profiler * timers)
{
  float (*restrict f) __attribute__ ((aligned (64))) = (float (*)) f_vec->data;
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]]) u_vec->data;
  float (*restrict ure)[ure_vec->size[1]][ure_vec->size[2]] __attribute__ ((aligned (64))) = (float (*)[ure_vec->size[1]][ure_vec->size[2]]) ure_vec->data;

  /* Flush denormal numbers to zero in hardware */
  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);
  _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);

  for (int time = time_m, t0 = (time)%(2), t1 = (time + 1)%(2); time <= time_M; time += 1, t0 = (time)%(2), t1 = (time + 1)%(2))
  {
    /* Begin section0 */
    START_TIMER(section0)
    for (int x = x_m; x <= x_M; x += 1)
    {
      #pragma omp simd aligned(u:64)
      for (int y = y_m; y <= y_M; y += 1)
      {
        u[t1][x + 1][y + 1] = u[t0][x + 1][y + 1] + 1;
      }
    }
    STOP_TIMER(section0,timers)
    /* End section0 */
  }

  /* Begin section1 */
  START_TIMER(section1)
  for (int freq_dim = freq_dim_m; freq_dim <= freq_dim_M; freq_dim += 1)
  {
    for (int time = time_m, t0 = (time)%(2); time <= time_M; time += 1, t0 = (time)%(2))
    {
      if (time >= f[freq_dim])
      {
        for (int x = x_m; x <= x_M; x += 1)
        {
          #pragma omp simd aligned(u,ure:64)
          for (int y = y_m; y <= y_M; y += 1)
          {
            ure[freq_dim][x + 1][y + 1] = u[t0][x + 1][y + 1] + ure[freq_dim][x + 1][y + 1];
          }
        }
      }
    }
  }
  STOP_TIMER(section1,timers)
  /* End section1 */

  return 0;
}
```

### PR DESCRIPTION
**Pull request by FabioLuporini**
fixes #2007

### CODE REVIEW

File: devito/passes/clusters/blocking.py [hunk 1]
```diff
@@ -184,6 +184,10 @@ def callback(self, clusters, prefix):
             return clusters
 
         d = prefix[-1].dim
+        if d.is_Default:
```
Comment (mloubout) [inline devito/passes/clusters/blocking.py:187]: Is this related? 

Not sure about the "small" but I get the idea and makes code nicer avoid weird blocks so I like it.

**Review by mloubout — APPROVED** [review]

File: devito/passes/clusters/blocking.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/passes/clusters/blocking.py:187]: no unrelated sorry forgot to mention!

Comment (mloubout) [reply devito/passes/clusters/blocking.py:187]: Ok tanks. Still happy with it, GTG to be merged once green.

File: devito/ir/clusters/algorithms.py [hunk 2]
```diff
@@ -198,8 +199,17 @@ def guard(clusters):
                 else:
                     d = cd.parent
 
+                # If `cd` uses, as condition, an arbitrary SymPy expression, then
+                # we must ensure to nest it inside the last of the Dimensions
+                # appearing in `cd.condition`
+                if cd._factor is not None:
+                    k = d
+                else:
+                    dims = pull_dims(cd.condition)
```
Comment (georgebisbas) [inline devito/ir/clusters/algorithms.py:208]: Here, before pulling, we can evaluate the relation, cuz it could be reduced

Comment (georgebisbas) [reply devito/ir/clusters/algorithms.py:208]: No, it is late to optimize here. SHould happen way earlier at construction time of ConditionalDimension

**Review by georgebisbas — APPROVED** [review]
Since my comment is totally orthogonal to this PR,  I am approving this. Thanks.
